Fixed #214 - Added support for `s3a` scheme

pavibhai opened the pull request:
**Summary**
* Added support for s3a scheme when referring to paths on AWS S3
* scheme is ignored when comparing s3 and s3a paths for e.g. in overlap checks
* Tests have been added
* Some tests have been modified to run parameterized both for `s3` and `s3a`

**Motivation**
In the absence of this change, for tables involving `S3AFileSystem` the default scheme of `s3a` was used in the Iceberg metadata files. When you try to register this table it fails as `s3a` is not a valid scheme in Apache Polaris

As a workaround we had to at a minimum update the snapshot file to change the table location from `s3a://` to `s3://` for the table to be registered with Polaris

**Desired Behavior**
We should be able to treat both `s3://` and `s3a://` as valid AWS S3 paths

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/aws/S3Location.java:
```diff
@@ -0,0 +1,60 @@
+public class S3Location extends StorageLocation {
+  private static final Pattern URI_PATTERN = Pattern.compile("^(s3a?)://(.+)$");
+  private final String scheme;
+  private final String object_key_w_bucket;
```
I think a pass for style would be useful

pavibhai replied:
Yeah, missed that one, will change the variable name.

pavibhai replied:
Fixed

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -32,6 +33,8 @@ public static StorageLocation of(String location) {
     // TODO implement StorageLocation for all supported file systems and add isValidLocation
     if (AzureLocation.isAzureLocation(location)) {
       return new AzureLocation(location);
+    } else if (S3Location.isS3Location(location)) {
```
This looks like exactly the right approach to me

eric-maynard commented on plugins/spark/v3.5/integration/src/intTest/java/org/apache/polaris/spark/quarkus/it/SparkIT.java:
```diff
@@ -30,175 +30,190 @@
-    sql("DROP TABLE %s", newIcebergTable);
-    sql("DROP NAMESPACE %s", namespace);
+  @QuarkusIntegrationTest
+  static class S3ATest extends BaseTest {
```
Ah, yeah, found it. Let's not duplicate all these tests, that's quite expensive. Can we just do one smoke test reading/writing a table?

pavibhai replied:
Sure that makes sense. Let me add one test that does the basic and cross check

pavibhai replied:
I have reverted this and added a new test `PolarisS3InterOperabilityTest.java`

eric-maynard commented on plugins/spark/v3.5/integration/src/intTest/java/org/apache/polaris/spark/quarkus/it/SparkCatalogPolarisIT.java:
```diff
@@ -20,5 +20,18 @@
-public class SparkCatalogPolarisIT extends SparkCatalogBaseIT {}
+public class SparkCatalogPolarisIT {
+  @QuarkusIntegrationTest
+  static class S3ATest extends SparkCatalogBaseIT {
```
Same comment -- these tests are slow, let's not duplicate all of them

eric-maynard commented on plugins/spark/v3.5/integration/src/intTest/java/org/apache/polaris/spark/quarkus/it/SparkCatalogIcebergIT.java:
```diff
@@ -21,31 +21,47 @@
+  }
+
+  @QuarkusIntegrationTest
+  static class S3ATest extends BaseTest {
```
ditto

pavibhai replied:
I have reverted this and added a new test `PolarisS3InterOperabilityTest.java`

eric-maynard commented on runtime/service/src/test/java/org/apache/polaris/service/quarkus/admin/PolarisOverlappingCatalogTest.java:
```diff
@@ -112,6 +128,20 @@ public void testBasicOverlappingCatalogs(boolean initiallyExternal, boolean late
 
+  @ParameterizedTest
+  @CsvSource({"s3,s3a", "s3a,s3"})
+  public void testBasicOverlappingCatalogWSchemeChange(String rootScheme, String overlapScheme) {
```
Great idea to test this

eric-maynard reviewed (commented):
Looks good overall, I left some comments on tests. You may need to make some adjustments to make this compatible with #1686 but I otherwise feel this is pretty ready to merge

flyrain commented on polaris-core/src/test/java/org/apache/polaris/service/storage/aws/AwsCredentialsStorageIntegrationTest.java:
```diff
@@ -79,7 +80,7 @@ public void testGetSubscopedCreds() {
               return ASSUME_ROLE_RESPONSE;
             });
-    String warehouseDir = "s3://bucket/path/to/warehouse";
+    String warehouseDir = scheme + "://bucket/path/to/warehouse";
```
Can we test the mixed use cases? For example, we create a catalog with allowed locations like `s3://bucket1/abc/`, while some of tables paths are with `s3a`. Does credential vending work in that case? Also i'm not sure how much this simulation can catch. It'd be nice to have a test against the real s3 storage for the mixed use cases.

pavibhai replied:
> Does credential vending work in that case? Also i'm not sure how much this simulation can catch. It'd be nice to have a test against the real s3 storage for the mixed use cases.

@flyrain Yes from my integration test it works, but your question is about a test in the project itself. 
Do we have any means for performing integration tests with AWS services already in the project? Otherwise this is something we should discuss.

flyrain replied:
The regression tests under dir `regtests` are designed for that. It only run against local file system now. S3, Azure, and GCP tests are skipped. @jbonofre and me are working on an AWS account to enable that. Here is the command to run them. You will see s3 ones are skipped
```
./gradlew run
env POLARIS_HOST=localhost ./regtests/run.sh
```

pavibhai replied:
thanks @flyrain will check that out

pavibhai replied:
@flyrain I have adjusted the test in `AwsCredentialsStorageIntegrationTest:testGetSubscopedCreds` to validate that the policy that we generate does not have any references to S3A.

This should be sufficient proof on this until we get to the test in `regtests`

pavibhai replied:
> Can we test the mixed use cases? For example, we create a catalog with allowed locations like `s3://bucket1/abc/`, while some of tables paths are with `s3a`. Does credential vending work in that case? Also i'm not sure how much this simulation can catch. It'd be nice to have a test against the real s3 storage for the mixed use cases.

These tests have been moved to `PolarisS3InterOperabilityTest`, please check

flyrain reviewed (commented):
Thanks a lot for working on it @pavibhai !  The approach looks good to me. And thanks for adding a lot of tests. Left some comments.

pavibhai commented on polaris-core/src/main/java/org/apache/polaris/core/storage/aws/S3Location.java:
sure

pavibhai commented on polaris-core/src/test/java/org/apache/polaris/service/storage/StorageLocationTest.java:
```diff
@@ -77,4 +81,12 @@ public void testSpecialCharacters() {
+
+  @ParameterizedTest
+  @CsvSource({"s3,s3a", "s3a,s3"})
+  public void testPrefixValidationIgnoresScheme(String parentScheme, String childScheme) {
```
Moved this

pavibhai commented:
> Looks good overall, I left some comments on tests. You may need to make some adjustments to make this compatible with #1686 but I otherwise feel this is pretty ready to merge

I have rebased onto the current version. I can submit another PR to handle the `withoutScheme` method better in S3Location as that information is already available with the regex parsing.

pavibhai commented:
@eric-maynard and @flyrain thanks for your comments, here are the key highlights of the change
* Rebased onto main (force push)
* Added a test `PolarisS3InterOperabilityTest` to include all of the S3 and S3A interoperability tests and removed the changes onto the IT tests
* Added a test `S3LocationTest` that includes the unit tests on S3Location
* Fixed a test to show case that even with S3A the policy content generated has no reference to S3A so should work similar to S3 for vending credentials

eric-maynard commented on runtime/service/src/test/java/org/apache/polaris/service/quarkus/admin/PolarisS3InterOperabilityTest.java:
`Interoperability` is one word

eric-maynard reviewed (commented):
The tests look way better now, thanks for the update!

pavibhai commented on polaris-core/src/test/java/org/apache/polaris/core/storage/aws/S3LocationTest.java:
```diff
@@ -0,0 +1,49 @@
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class S3LocationTest {
```
Added

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/aws/S3Location.java:
```diff
@@ -0,0 +1,69 @@
+public class S3Location extends StorageLocation {
+  private static final Pattern URI_PATTERN = Pattern.compile("^(s3a?)://(.+)$");
+  private final String scheme;
+  private final String objectKeyWBucket;
```
Is this always the same as `withoutScheme`? If so, let's remove it and use that. If not, let's improve the name please

eric-maynard reviewed (commented):
The tests look great now; just one comment on `S3Location` and this LGTM

pavibhai commented on runtime/service/src/test/java/org/apache/polaris/service/quarkus/entity/CatalogEntityTest.java:
```diff
@@ -171,15 +179,29 @@ public void testValidAllowedLocationPrefix() {
             .setStorageType(StorageConfigInfo.StorageTypeEnum.AZURE)
             .setTenantId("tenantId")
             .build();
+    CatalogProperties prop = new CatalogProperties(basedLocation);
```
done

flyrain commented:
Thanks @pavibhai for working on it. Thanks @eric-maynard for the review!
